### PR TITLE
[FIX] hr_expense: add missing setup call on list renderer

### PR DESCRIPTION
**Pull request by ushyme**
Currently, if you open any list view in `hr_expense` and attempt to insert it into a spreadsheet, nothing happens.

### Steps to reproduce

* install `hr_expense`
* open a list view
* click "Favorites > Insert list in spreadsheet" (or "Favorites > Link menu in spreadsheet")

You should see that nothing happens.

opw-3633963

### CODE REVIEW

**Comment by rugo-odoo** [comment]
I don't understand how this fixes the problem (or anything). Isn't that the default behavior ?

**Comment by ushyme** [comment]
TLDR: `patch` order matters.


### Cause

To clearly explain this issue, we first have to understand one thing about the v16 implementation of the `patch` function: When it's used to override methods, it introduces a `_super` property to allow the new method to reference the original method it overrides. Crucially, this reference is fixed at the time of patching; it doesn't dynamically update if the method it points to is later overridden by another patch.


Now consider the following test scenario. It's a minimal setup to reproduce the issue at hand:

```javascript
QUnit.debug("path order test", async function (assert) {
    class A {
        setup() {
            assert.step("original");
        }
    }

    class B extends A {
        setup() {
            super.setup()
        }
    }

    patch(B.prototype, "", {
        setup() {
            assert.step("b patch");
            this._super();
        }
    });

    patch(A.prototype, "", {
        setup() {
            assert.step("a patch");
            this._super();
        }
    });

    const b = new B();

    b.setup();
    assert.verifySteps([
        "b patch",
        "a patch",
        "original"
    ]);

});
```
If you try to run this test, it will fail. However, it will pass if you reverse the order of the patches. 
The issue comes from the static nature of the `_super` reference established by the patch function. When the patch is applied to `B.prototype.setup`, `_super` points to `A`'s setup method as it existed at that moment. Subsequent patches to `A.prototype.setup` do not update `_super` in `B`'s patch, causing the "a patch" to be bypassed.



### Fix

To resolve this, a simple yet effective change is made: explicitly invoking the superclass method within `B`'s `setup` method using `super.setup()`. This direct call ensures that the most current version of `A`'s setup method is executed, regardless of the patching sequence.

```javascipt
class B extends A { 
    setup() {
        super.setup();
    }
}

**Review by rugo-odoo — APPROVED** [review]
LGTM (cc @william-andre)

**Review by william-andre — COMMENTED** [review]
Yeah, I need an explanation on this one...

**Comment by ushyme** [comment]
TLDR: `patch` order matters.


### Cause

To clearly explain this issue, we first have to understand one thing about the v16 implementation of the `patch` function: When it's used to override methods, it introduces a `_super` property to allow the new method to reference the original method it overrides. Crucially, this reference is fixed at the time of patching; it doesn't dynamically update if the method it points to is later overridden by another patch.


Now consider the following test scenario. It's a minimal setup to reproduce the issue at hand:

```javascript
QUnit.debug("path order test", async function (assert) {
    class A {
        setup() {
            assert.step("original");
        }
    }

    class B extends A {
        setup() {
            super.setup()
        }
    }

    patch(B.prototype, "", {
        setup() {
            assert.step("b patch");
            this._super();
        }
    });

    patch(A.prototype, "", {
        setup() {
            assert.step("a patch");
            this._super();
        }
    });

    const b = new B();

    b.setup();
    assert.verifySteps([
        "b patch",
        "a patch",
        "original"
    ]);

});
```
If you try to run this test, it will fail. However, it will pass if you reverse the order of the patches. 
The issue comes from the static nature of the `_super` reference established by the patch function. When the patch is applied to `B.prototype.setup`, `_super` points to `A`'s setup method as it existed at that moment. Subsequent patches to `A.prototype.setup` do not update `_super` in `B`'s patch, causing the "a patch" to be bypassed.


This simplified version of the problem, maps to the real problem, like so:
* `A` is `ListRenderer`
* `B` is `ExpenseListRenderer`
* `'a patch'` is `'knowledge_list_embeddable'`
* `'b patch'` is `'expense_list_renderer_qrcode'` or `'expense_list_renderer_qrcode_dzone'`

### Fix

There are several approaches to address this issue. One option is to consider a potential rewrite of the patch function, although this is beyond my current scope, and it may cause problems elsewhere.

Another potential solution is to attempt to enforce the correct order of the problematic patches. However, this would be difficult since these patches are located in two unrelated and independent modules.

Furthermore, it is worth noting that this issue has been resolved in version 17, where mixins are employed instead of patches.


Given these considerations, I have opted for this simple change. This direct call ensures that the most current version of `A`'s setup method is executed, regardless of the patching sequence.

```javascript
class B extends A { 
    setup() {
        super.setup();
    }
}
```
When `setup()` is now called on an instance of `B`, it first executes the patched "b patch" method. Inside this method, `super.setup()` directly invokes the current setup method of `A`, which, due to the later patch, is the "a patch" version. This method then calls its own `_super`, referring to the original `setup` method of `A`, thus completing the intended sequence: "b patch", "a patch", and finally "original".

**Comment by rugo-odoo** [comment]
It won't be a bad idea to add a comment directly on the code.

**Review by william-andre — APPROVED** [review]
@robodoo r+


Good that it is fixed in 17.0 then
@fw-bot up to saas-16.4
